### DISCUSSION TITLE
Cart Block: Filtering "shipping amount" and "total amount"

### DISCUSSION THREAD

**Discussion by larsenlarsson** [woocommerce/woocommerce#44999] [WooCommerce, Blocks, and Site Editing]
Dear WooCommerce community, 
we hope you are doing well and maybe you have some advice for us.

In the cart, which is included via shortcode, we added an output for the total amount of shipping costs with the hook `woocommerce_cart_shipping_method_full_label` and changed the output of the total amount with `woocommerce_cart_totals_after_order_total` to add/change tax information:

![cart-shortcode](https://github.com/woocommerce/woocommerce-blocks/assets/2648926/9549ffb3-d03c-43b0-bb78-0cdbbab727bb)

These hooks are not available in the cart block.

With the help of the documentation https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/third-party-developers/extensibility/checkout-block/available-filters.md we managed to filter the cartItemPrice (see https://github.com/woocommerce/woocommerce-blocks/discussions/10700). But for the places shown in my screenshot there is no way in the documentation to change the output. 

Do you have a hint for us how we can influence the output of these data in the cart block?

Thanks in advance for any help.

**Comment by opr** [comment]
Hi @larsenlarsson, good question!

My suggestion won't get the exact same look and feel as what you're originally doing, but did you consider [adding a Slot/Fill](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/third-party-developers/extensibility/checkout-block/available-slot-fills.md#experimentalordershippingpackages) under the shipping row (https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/third-party-developers/extensibility/checkout-block/available-slot-fills.md#experimentalordershippingpackages) to add these items?

We also have a slot fill that appears below the total where you might want to add your breakdown too. https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/third-party-developers/extensibility/checkout-block/available-slot-fills.md#experimentalordermeta

**Comment by larsenlarsson** [comment]
Hi @opr ,

thanks for your answer and your time.

> but did you consider [adding a Slot/Fill](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/third-party-developers/extensibility/checkout-block/available-slot-fills.md#experimentalordershippingpackages) under the shipping row

This would not meet the legal requirements that the tax info must appear directly after every shipping cost, especially when there is more than one flat rate:
![shipping_costs](https://github.com/woocommerce/woocommerce-blocks/assets/2648926/e194f1fe-5658-4a7a-8454-6f76bb45d681)

===

> We also have a slot fill that appears below the total where you might want to add your breakdown too.

This would be possible, but we would then need to be able to hide the taxes shown by default in the block. 
![hide_taxes_total](https://github.com/woocommerce/woocommerce-blocks/assets/2648926/ad171045-06a2-481c-a4c7-6c54188ab749)

Would this be possible and if so, how?

Furthermore, it would have to be ensured that no other slot appears between the price and our slot. Something like "priority" does not exists for "registerCheckoutFilters", right?

Any other suggestions would be greatly appreciated.

Would it be possible that you could extend the filters (https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/third-party-developers/extensibility/checkout-block/available-filters.md) for any output?

Thanks in advance for your efforts.

**Reply by opr** [reply]
> This would not meet the legal requirements that the tax info must appear directly after every shipping cost, especially when there is more than one flat rate:

Ahh, ok, I see, then that won't work. Out of curiosity, what country/state are you from that makes this requirement? If it's something that a lot of merchants are going to need, we may want to look into adding it as a toggle option on the Cart/Checkout blocks.

> Furthermore, it would have to be ensured that no other slot appears between the price and our slot. Something like "priority" does not exists for "registerCheckoutFilters", right?

Yes, this is unfortunately true, there's no priority system in place for this. That's a limit imposed by the Slot/Fill system added by Gutenberg, and we can't change that.

> Would it be possible that you could extend the filters ([trunk/docs/third-party-developers/extensibility/checkout-block/available-filters.md](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/third-party-developers/extensibility/checkout-block/available-filters.md)) for any output?

This isn't something we'd be able to do, that would mean adding a filter for every output and it would add a maintenance burden that we can't possibly take on.

I think the best next step to look at would be to look whether we should add an option to break down the tax rates per shipping item too. cc @pmcpinto

**Reply by larsenlarsson** [reply]
Hi @opr, thanks again for your time and your response.

> what country/state are you from that makes this requirement?

Our plugin users are shop admins from Switzerland and the EU, significantly from Austria and Germany.

According to legal regulations, which we have legally checked, among other things, the tax information should appear in the cart and checkout for all item prices, as well as for shipping and fees, directly after the price, as well as after the total amounts.

We display the tax amount, tax name and tax rate. Some shop admins would like to see a general output like "Incl. tax" instead of the split output by tax rates (as seen my screenshots). Where the output can be customized regarding the wording ("incl. tax", "Incl. vat", "inclusive Taxes", "Inclusive VAT" , "Contains"). However, most shop admins prefer the itemized output per included tax rate consisting of one word for "including" followed by tax amount, tax name, tax rate.

> I think the best next step to look at would be to look whether we should add an option to break down the tax rates per shipping item too

With fees, this should also be possible.

===

It would be great if the tax output for third party developers could be filterable / customizable. Do you see a way that this can be achieved? That way we could customize the legal requirements and the shop admins wishes as before in shortcode cart and shortcode checkout, at least for the tax rate output.

If you have any alternative ideas or suggestions, we would be grateful.

Best regards and have a nice weekend :)

**Comment by pmcpinto** [comment]
> I think the best next step to look at would be to look whether we should add an option to break down the tax rates per shipping item too. cc @pmcpinto

Do you mean having a wp-admin setting that allows users to change how Taxes are displayed in the cart and checkout?

**Reply by opr** [reply]
Yes, but I meant specifically on the Cart/Checkout blocks. Or we could look into extensibility points to make this more free form, but I feel like if we do that we will just end up implementing more and more filters.

**Reply by zyra87** [reply]
> but I feel like if we do that we will just end up implementing more and more filters.

And that would be fine, because everything should be customizable as in the shortcode cart and checkout pages.

**Comment by larsenlarsson** [comment]
Hey folks, hi @opr 

I would like to briefly inform you that we have found a solution. 
**The solution is anything but elegant, but it works.**

We have extended the Store API so that we can access in our extension the necessary strings for the taxes of each available shipping method.

Now comes the inelegant part: we add a component in the "`ExperimentalOrderMeta`" slot, which gets a <style>-block containing "_styles_" like this:
`span#radio-control-0-flat_rate\:7__description:after{ content: ' Includes $0.40 VAT (19%)'; }`

The result in the cart  block  is as we expected:

<img width="427" alt="ugly_solution" src="https://github.com/woocommerce/woocommerce-blocks/assets/2648926/38c4b3be-d0e1-443d-9485-8f382cd1af00">

It is not very nicely programmed, but the result is as we need it.
